### PR TITLE
Tests setup

### PR DESCRIPTION
**Pull request by MunGell**
Hey guys,

I was using tests to make sure pagination limits are working fine, so this PR contains my test setup in case you will find it useful. I've never done testing of React components before, so I was mostly relying on what other repositories use for these purposes.

I've left simple pagination test in this branch as an example of how tests work.

Please consider this PR experimental and feel free to decline if it does not match your criteria.

### CODE REVIEW

**Comment by beni55** [comment]
Please consider this PR experimental and feel free to decline if it does not work for you.

**Comment by JedWatson** [comment]
@MunGell thanks - I really appreciate you doing the work to start getting Elemental under test. I'm (almost) happy to merge this as-is, two things first;

1. You're including `jsdom@2.0.0`, which doesn't work with node.js 4.0. The latest version _does_ work with 4.0 but isn't backwards compatible w/ node 0.x. Bit of a dilemma here now, as it creates a hard lock-out for development if someone is on the wrong version of node. I'm tempted to solve this by at least implementing the latest version (which has breaking API changes and will require a small refactor)

2. What do you think of using [legit-tests](https://github.com/Legitcode/tests)? @zackify did a good write-up here: http://reactjsnews.com/testing-the-easy-way/

It might be worth looking at how simple it is to convert these over to the latest jsdom and legit-tests. If you're happy to do this let me know and I'll hold off on merging this PR :)

**Comment by zackify** [comment]
@JedWatson thanks for mentioning legit tests:) If you opt in and use it, the library will depend on the latest jsdom already, but I'm sure you know that from the writeup. Let me know if you guys have any questions

**Comment by nkbt** [comment]
https://medium.com/javascript-scene/why-i-use-tape-instead-of-mocha-so-should-you-6aa105d8eaf4

**Comment by MunGell** [comment]
@JedWatson I went through the article and the source code of `legit-test`, it looks very interesting.
I can try to implement it here, but cannot guarantee any "ETA".

@nkbt Thank you for the link. I like `Tape`'s approach, however I don't agree with the author that it is the only right solution for testing. The whole comparison looks controversial to me.

Although, Tape might be superior to Mocha/Chai, in my opinion we should also consider popularity of tools when implementing them in a project with (possibly) many collaborators. Tests are not something people naturally love to do and less popular framework could be additional barrier for writing them.

**Comment by nkbt** [comment]
@MunGell for my personal projects and libs I use Karma + Jasmine + inject-loader (for deps mocking) + Esparta (for ES6 code coverage), but for some production projects and other OSS - tape. And I think that eventually I'd love to move to Tape. Simple with zero overhead and great compatibility with other tools.

Regarding PR I'm not sure `jsdom` is the best way to test React components. For most components it is sufficient to go with DOM-less approach: http://simonsmith.io/unit-testing-react-components-without-a-dom/ and sometimes you actually want to render component and test it in more-or-less real environment with `componentWillUnmount`, so `TestUtils.renderIntoDocument` is a great option.

I mean for something like:
```js
const div = document.createElement('div');
React.render(<PageClick onClick={() => {}}><span>Test</span></PageClick>, div);
React.unmountComponentAtNode(div);
expect(global.window.removeEventListener).toHaveBeenCalled();
```

You can check https://github.com/nkbt/react-component-template for inspiration/tooling

File: test/utils.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import React from 'react/addons'
+
+var TestUtils = React.addons.TestUtils;
+
+export default {
+
+	getComponent: function (Component, props) {
```
Comment (nkbt) [inline test/utils.js]: I'm not convinced these helpers are needed. The barely do any useful work, just adding extra conventions.

**Comment by JedWatson** [comment]
> Tests are not something people naturally love to do and less popular framework could be additional barrier for writing them

:+1: I agree that the easiest / most readily understandable framework for contributors to add tests is the best one to have.

(no judgements on any of the frameworks mentioned in that statement, just agreeing with the sentiment)

**Comment by JedWatson** [comment]
More on my comment above, TBH usually when I go to write tests in a project (that isn't mine) I look at the existing tests and copy the style / features used. So if it's not a library I'm familiar with (or there are a bunch of custom helpers) that's usually OK as long as there's documentation on how to use what's there & plenty of good examples (read: existing tests ;)

I'd love to see a test-off between this PR, `tape` and `legit-test`... might be educational. @nkbt @zackify you guys interested?

**Comment by nkbt** [comment]
@JedWatson that's a good idea, I'll PR a Tape alternative tomorrow for comparison.

File: test/components/pagination.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import utils from '../utils';
+
+import Pagination from '../../lib/components/Pagination';
+
+describe('Component: Pagination', function () {
+
+	describe('Feature: Basics', function () {
+
+		it('renders correct number of pages', function () {
+			var props = {
```
Comment (nkbt) [inline test/components/pagination.js]: here and further - `const`

File: test/components/pagination.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
+import { expect } from 'chai';
+import utils from '../utils';
+
+import Pagination from '../../lib/components/Pagination';
+
+describe('Component: Pagination', function () {
```
Comment (nkbt) [inline test/components/pagination.js]: can't see why not omit `function` and simply write `() => {` ;)

**Comment by JedWatson** [comment]
@nkbt awesome!

File: test/components/pagination.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+				list = utils.getDOMNodeWithClass(Pagination, props, 'Pagination__list');
+
+			expect(Array.from(list.children, utils.getInnerHTML)).to.eql(['1', '2', '3', '4', '5']);
+		});
+
+	});
+
+	describe('Feature: Limit', function () {
+
+		let getPages = function (props) {
```
Comment (nkbt) [inline test/components/pagination.js]:
```js
const getPages = props => Array.from(utils.getDOMNodeWithClass(Pagination, props, 'Pagination__list').children, utils.getInnerHTML)
```

Comment (zackify) [inline test/components/pagination.js]: I'll write a PR with legit tests;) The only thing I can say is that it does
support shallow rendering and the big thing for me is it makes your tests
readable. You can easily figure out what is being tested, I feel like that
does not create a higher barrier of entry, but a lower one.

**Comment by nkbt** [comment]
By the way, @MunGell, @zackify, please include code coverage reporting too. At the moment current PR lacks coverage report.

**Comment by MunGell** [comment]
Looking forward to see other implementations, indeed it will be educational to see different approaches :)

**Comment by zackify** [comment]
Anyone interested can take a look at #63

**Comment by ericelliott** [comment]
Hi guys. I have a fair amount of experience unit testing React components. I'm writing a book on Universal JavaScript with React, so I've tried most of the popular choices.

Tape is hands down -- no contest -- my favorite test runner. Read [Why I Use Tape Instead of Mocha & So Should You](https://medium.com/javascript-scene/why-i-use-tape-instead-of-mocha-so-should-you-6aa105d8eaf4) to learn why.

As for testing React components, the existing test utils are a bit of a mess when it comes to API design. I've been using Cheerio instead, but I haven't settled on a really great solution for testing synthetic events... 

... but I just learned about a new alternative today: [teaspoon](https://github.com/jquense/teaspoon). I'm very hopeful that it will be a great solution. =)

**Comment by nkbt** [comment]
@ericelliot yeah, there still is #64 here :) no teaspoon though. Or enzyme.

**Comment by ericelliott** [comment]
@nkbt Try replacing those utils with teaspon, and using JSX right in the tests. You'll get much more readable results. =)

**Comment by MunGell** [comment]
@JedWatson should we close all test-related PRs since the decision was made?

**Comment by ericelliott** [comment]
:+1:

**Comment by nkbt** [comment]
I agree

**Comment by MunGell** [comment]
@JedWatson @jossmac guys, I am closing my PRs as they are not relevant any more

**Comment by JedWatson** [comment]
Thanks for your contributions @MunGell - even though we ended up going with #64 I really appreciate your work here and in #96!

**Comment by JedWatson** [comment]
(where by #64 I mean #116!)

**Pull request closed without merging**